woocommerce/woocommerce-ios
# Products: Add duplicate product feature

Closes: #7255 


### Description

This PR adds the feature to duplicate products.

This feature is available for all products that have been saved on the server, regardless of their status (published/draft).

This feature is implemented on WooCommerce Core [here](https://github.com/woocommerce/woocommerce/blob/5d7f6acbcb387f1d51d51305bf949d07fa3c4b08/includes/admin/class-wc-admin-duplicate-product.php#L97), but not available on the API.

The idea is simply to create a copy of the original product, rename it, set its status to draft, and make sure to clear the SKU value since it needs to be unique. The implementation on Core handles generating a new unique value based on the original one, but this requires checking for existing values which may take time, so I'm just resetting it to nil for simplicity.

For variable products, the original product's variations are also duplicated to the new one.

### Testing instructions

- Navigate to the Products tab, and select a simple product with SKU value. 
- Tap the "..." button on the top right and select Duplicate. Notice in Xcode console: `🔵 Tracked product_detail_duplicate_button_tapped`
-  Notice that after the process completes, the view is reloaded with the copied product, with the name updated and no SKU value. The status of the product should be Draft. Most of the other values should be the same as the original one, except for reviews, rating, slug, and product ID. Xcode console should log: `🔵 Tracked duplicate_product_success`
- Repeat the test with a variable product. After duplicating, the new product should have the same variations as the original.
- Turn off the Internet connection and try duplicating product again. Xcode console should log `🔵 Tracked duplicate_product_failed` with error information in the properties.

### Screenshots


https://user-images.githubusercontent.com/5533851/190574415-498989a5-e2b3-4e83-928e-94d8397c70d2.mp4



---
